### ISSUE TITLE
SyncBatchNorm size check

### ISSUE THREAD

**Issue by elmirador** [pytorch/pytorch#36865]
https://github.com/pytorch/pytorch/blob/9c7e604c607e6d9051af888ede9a009be2bda676/torch/nn/modules/_functions.py#L12-L13

Shouldn't the size check be `size * world_size == 1`?
Even if I'm using one sample per batch, across all GPUs (if using SyncBatchNorm) I'll have enough data to calculate the statistics.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by pritamdamania87** [comment]
This does make sense and maybe the check should be `count_all == 1`? Looks like this check was added in https://github.com/pytorch/pytorch/pull/29626, @albanD I was wondering if you had thoughts on the proposal here given you reviewed that PR?

**Comment by albanD** [comment]
What I would like to check is that `torch.batch_norm_stats()` handle a single input. And won't return inf/nan (or other bad stuff). Because that would pollute all the computed stats.
Unfortunately that function is not documented...
If it does handle it properly then yes we should change that check to count_all.
If it does not handle it, we need to change the strategy to accumulate the std that handle subsets of size 1.

**Comment by elmirador** [comment]
> What I would like to check is that `torch.batch_norm_stats()` handle a single input. And won't return inf/nan (or other bad stuff). Because that would pollute all the computed stats.
> Unfortunately that function is not documented...
> If it does handle it properly then yes we should change that check to count_all.
> If it does not handle it, we need to change the strategy to accumulate the std that handle subsets of size 1 and then change that check.

The corresponding function of `torch.batch_norm_stats()` is `batch_norm_stats_cuda_template()` at `aten/src/ATen/native/cuda/Normalization.cuh:L634`.
I did a quick check on the source code. The template function invokes CUDA kernel `batch_norm_collect_statistics_kernel()`.
The kernel function returns mean and **biased** variance of the input(L298, `var_n / N`), instead of an unbiased one where you need to `var_n / (N - 1)`. Thus no problem when `N == 1`.

The reproduction code below runs without NaN or inf. `mean` is equal to `a` and var is an all-zero tensor.
`
import torch
import numpy as np

a=torch.tensor(np.random.random([1,32,1,1])).cuda()
mean,var=torch.batch_norm_stats(a, 0.0) # input and eps

print (a.squeeze())
print (mean)
print (var)
`

**Comment by albanD** [comment]
Ok then it sounds like it would work fine just to change that condition (and add the corresponding test).
We'll be happy to accept a PR for this.
